### DISCUSSION TITLE
G7 Companion mode sometimes out of sync with double readings.

### DISCUSSION THREAD

**Discussion by grahamjenson** [NightscoutFoundation/xDrip#2767] [General]
I used to use companion mode with CamAPS and it worked flawlessly. However, with the dexcom app the timing of readings gets out of sync, so readings can take many minutes to actually be read by xdrip and sometimes the same reading is read twice. This is not a giant issue, though it does delay any AndroidAPS decisions. 

As a short term workaround I now just use dexcom share as the BG source, but would like companion mode especially since it doesnt rely on internet connection.

I am not sure the difference between how dexcom sends out notifs vs CamAPS, but I did notice more deduplication period code in UIBasedCollector.java class. This might be a cause.

**Comment by frendmatt** [comment]
I noticed improvements in companion mode when enabling **settings** >  **less common settings** > **Bluetooth settings** > **companion bluetooth**

Hope this helps

**Comment by grahamjenson** [comment]
@frendmatt That seems to have worked. Cheers for that. But why and how does that work? Should this be on by default for all companion collectors?

**Reply by frendmatt** [reply]
Not sure, I haven't looked in the code. I just assumed that it would scan for when the G7 connects to the device for the dexcom app and use that to know when to pull from the notification. I tried scanning through the repo quick, but I'm on my phone and couldn't find where any of the companion mode files sat

**Reply by grahamjenson** [reply]
UIBasedCollector is where the companion app notifications stuff is processed. The only direct call to BlueTails (which is where the option is used) is to initialize it.

I think the BlueTails checks BT connections and sets the `lastReadingTimestamp` which the UIBasedCollector uses to time the readings.

**Comment by jamorham** [comment]
Companion bluetooth doesn't work on all phones. Pretty sure it needs Android 12 and above and could cause problems sometimes on some handsets which is why it isn't enabled as default. It also only works with G6/G7. If it works on your handset though then it is a definite improvement and also provides back filling.

**Comment by Arakon** [comment]
Apologies, just realized this moved here. So I'll add this here too, for reference:

I've had no such issue on a Pixel 6a. However, when backfilling after a long signal loss, only 1 value every 40 minutes seems to be added, while the actual G7 app gets the full backfill.
This happens when backfilling several hours, if the backfill is only 30 mins or so, all data is there.

Signal loss at 3 am until 6 am:

![Screenshot_20230428-063412](https://user-images.githubusercontent.com/12016360/235157315-3475bbf0-6371-4993-831c-c4cd9abc90d4.png)
